### ISSUE TITLE
Regression. Failed to parse VCS url

### ISSUE THREAD

**Issue by Maks3w** [sarugaku/requirementslib#138]
With release `v1.4.1` Given a VCS entry like

`git+https://<token>:x-oauth-basic@github.com/<org>/<repo>.git@<git sha1>#egg=<package-name>`

```
Traceback (most recent call last):
  File "/usr/local/bin/pipfile2req", line 10, in <module>
    sys.exit(main())
  File "/usr/local/lib/python3.6/site-packages/pipfile_requirements.py", line 98, in main
    lines = convert_pipfile_or_lock(args.project, args.file, args.hashes, args.dev)
  File "/usr/local/lib/python3.6/site-packages/pipfile_requirements.py", line 93, in convert_pipfile_or_lock
    return _convert_pipfile_lock(full_path, hashes, dev)
  File "/usr/local/lib/python3.6/site-packages/pipfile_requirements.py", line 59, in _convert_pipfile_lock
    return lockfile.as_requirements(hashes, dev=dev)
  File "/usr/local/lib/python3.6/site-packages/requirementslib/models/lockfile.py", line 308, in as_requirements
    r = req.as_line(**kwargs)
  File "/usr/local/lib/python3.6/site-packages/requirementslib/models/requirements.py", line 2973, in as_line
    include_specifiers = True if self.specifiers else False
  File "/usr/local/lib/python3.6/site-packages/requirementslib/models/requirements.py", line 2762, in specifiers
    and (not isinstance(self.req, NamedRequirement) and self.req.setup_info)
  File "/usr/local/lib/python3.6/site-packages/requirementslib/models/requirements.py", line 2142, in setup_info
    ireq = self.parsed_line.ireq
  File "/usr/local/lib/python3.6/site-packages/requirementslib/models/requirements.py", line 1609, in parsed_line
    self._parsed_line = Line(self.line_part)
  File "/usr/local/lib/python3.6/site-packages/requirementslib/models/requirements.py", line 164, in __init__
    self.parse()
  File "/usr/local/lib/python3.6/site-packages/requirementslib/models/requirements.py", line 1123, in parse
    self.parse_link()
  File "/usr/local/lib/python3.6/site-packages/requirementslib/models/requirements.py", line 1017, in parse_link
    if (self.is_direct_url or vcs) and self.name is not None and vcs is not None:
  File "/usr/local/lib/python3.6/site-packages/requirementslib/models/requirements.py", line 570, in name
    if self.setup_info:
  File "/usr/local/lib/python3.6/site-packages/requirementslib/models/requirements.py", line 739, in setup_info
    self.setup_info = self.get_setup_info()
  File "/usr/local/lib/python3.6/site-packages/requirementslib/models/requirements.py", line 724, in get_setup_info
    setup_info = SetupInfo.from_ireq(self.ireq)
  File "/usr/local/lib/python3.6/site-packages/requirementslib/models/requirements.py", line 707, in ireq
    self.parse_ireq()
  File "/usr/local/lib/python3.6/site-packages/requirementslib/models/requirements.py", line 829, in parse_ireq
    if self.requirement is not None and self._ireq.req is None:
  File "/usr/local/lib/python3.6/site-packages/requirementslib/models/requirements.py", line 402, in requirement
    self.parse_requirement()
  File "/usr/local/lib/python3.6/site-packages/requirementslib/models/requirements.py", line 983, in parse_requirement
    "in the form {0}#egg=<package-name>.".format(url)
ValueError: pipenv requires an #egg fragment for version controlled dependencies. Please install remote dependency in the form git+https://<token>:x-oauth-basic@github.com/<org>/<repo>.git#egg=<package-name>.
```

**Comment by techalchemy** [comment]
Hm I’d definitely believe this. I made substantial changes to the parser to accommodate direct url parsing. My guess is whatever I’m doing to parse links in the `Line` class is dropping or malforming parts of the url. 

Feel free to have a look at this, or else I’ll get a fix together tomorrow and cut a follow up release as I’m currently preparing to release pipenv which has been held back by this for some time.  

Thanks for the report!

**Comment by techalchemy** [comment]
if you get a chance, it would be super helpful if you could test this with #139 (`pip install -e git+https://github.com/sarugaku/requirementslib.git@bugfix/138-vcs-url-parser#egg=requirementslib`) -- thanks again!

**Comment by Maks3w** [comment]
It does not throw the error.

Another issue I found is performance has dramatically dropped. It now takes half minute for to perform an operation than before it was almost instant.

**Comment by techalchemy** [comment]
@Maks3w that's because in the case of editable/vcs/file requirements, it now downloads the requirement and builds it in order to determine the dependency graph and the name.  I am debating whether to track this down and make it lazier and accept the supplied name (if it exists), or whether to consider it a feature.

I'm interested in feedback on this point -- what's your use case here? Just moving between formats (i.e. `requirements.txt` and `Pipfile` / `Pipfile.lock`)?

**Comment by techalchemy** [comment]
see #140 for the performance related issue

**Comment by Maks3w** [comment]
We lost our trust on Pipenv about to honor the versions declared on his Pipenv.lock file. So we have choose to use an hybrid system where Pipenv is responsible of create the lock file but then we move the results to a requirements file for to be installed with native pip command.

We gain reliability and speed of installation (50% less time)

For to be sure than requirements files are synced we use pipfile2req

```
dependency.tools:
	@pip install --no-cache-dir pipenv==2018.11.26 pipfile-requirements==0.1.0.post0 requirementslib==1.4.0

dependency.install: dependency.tools
	@pipenv install $(PACKAGE)
	@make dependency.lock

dependency.lock: dependency.tools
# It's not possible to use "hash" check because pip throws an exception for VCS repositories.
	@pipfile2req  > requirements/base.txt
	@pipfile2req --dev > requirements/dev.txt

dependency.check_sync: dependency.tools
	@pipfile2req | diff requirements/base.txt -
	@pipfile2req --dev | diff requirements/dev.txt -
```

**Comment by techalchemy** [comment]
@Maks3w pipenv should never ignore a lockfile's declared versions, but the issues you are running into may be around installing version-controlled dependencies in _non-editable_ mode -- if you don't use editable mode, the dependencies of the VCS requirement aren't locked in which case they may override the things in your lockfile, as pipenv prefers to install a working environment.

E.g. if you install 

`pipenv install git+https://github.com/requests/requests.git@master#egg=requests`

and then you also install

`pipenv install idna==2.7`

Pipenv will lock `idna==2.7` but wind up installing `idna==2.8` because that's what requests actually needs at runtime.  This is something I have fixed directly as a result of the changes I just merged in this project. Now, pipenv will always know what `git+https://github.com/requests/requests.git@master#egg=requests` requires and pass those requirements into the resolver before locking.

Resolving dependencies is complicated, and it's easy to simply say 'oh pipenv isn't respecting my lockfile' but it's doing pretty much the same thing as you, just prioritizing the VCS dependency's needs over whatever you are explicitly stating. If you have a concrete example beyond that though I'd be glad to take a look and get a fix up in the next release
